### DISCUSSION TITLE
posts/2022/05/using-dart-sass-hugo-github-actions-edition/

### DISCUSSION THREAD

**Discussion by giscus** [brycewray/comments#11] [Announcements]
# posts/2022/05/using-dart-sass-hugo-github-actions-edition/

Still more ways to make Hugo and Embedded Dart Sass work together.

https://www.brycewray.com/posts/2022/05/using-dart-sass-hugo-github-actions-edition/

**Comment by Studio384** [comment]
Is there any movement over at Hugo to just properly support Dart Sass out of the box? I can't really find anything in their GitHub repository, which is kinda nuts to me for a problem that has existed for quiet some time now.

**Reply by brycewray** [reply]
It’s a fair question! The short answer is: movement, yes, but a workable answer, no.

Here’s some history on the subject . . .

---

To be pedantic about it, I’ll note that Hugo actually **has** supported Dart Sass since [v.0.80.0](https://github.com/gohugoio/hugo/releases/tag/v0.80.0), released on 2020-12-31. At the time, there wasn’t yet a **stable** version of Embedded Dart Sass, but it was anticipated that such would appear within the first quarter of 2021. In fact, however, the first stable version actually wasn’t released  until February of **this** year.

In the period between Hugo v.0.80.0 and that first stable version of Embedded Dart Sass, questions like yours occasionally appeared, both in Hugo’s GitHub issues and on the Hugo Discourse forum. For example, on 2021-03-05, Hugo’s Bjørn Erik Pedersen [responded, in part](https://github.com/gohugoio/hugo/issues/8299#issuecomment-791245030), on GitHub:

> We cannot (I’m probably sure about that) embed this into Hugo’s binary.

... although he [subsequently mentioned](https://github.com/gohugoio/hugo/issues/8299#issuecomment-791274164) the possibility of making it part of a Homebrew “tap.” Apparently, that never came to fruition.

On 2022-02-11, Pedersen again acknowledged the issue when he [said this in response to a question on the Hugo Discourse forum](https://discourse.gohugo.io/t/using-dart-sass-hugo-and-netlify/37099/2):

> . . . even the NPM *[sic]* library currently requires an extra step to manually download the [Embedded Dart Sass] binary.

(This was as compared to the now-deprecated Libsass implementation, which apparently has existed as only a dependency within Hugo’s own binary.)

If you’re coming from the npm world, yes, this seems weird — just `npm install` something and it gets to be part of a `package.json` that you upload to your host, the host does its own `npm install` to pull down all the stuff in that `package.json`, and everything works. This situation is different. Exactly **why** it’s different, I lack the software chops to say; that might be something only Pedersen could answer.

---

I hope that answers your question, albeit in my as-usual overly wordy way. 😄

**Reply by Studio384** [reply]
So, the tl;dr is basically "this isn't going to be fixed"?

**Reply by brycewray** [reply]
> So, the tl;dr is basically "this isn't going to be fixed"?

Unless Pedersen comes up with a way to pack it all together that he hasn’t yet hinted at finding (which he did, early on, but I guess that didn’t turn out the way he’d hoped), apparently not.

**Comment by burtcrismore** [comment]
Thank you for all of the great documentation you've provided on the subject. I've been reading quite a bit about how to get Dart Sass running correct with Hugo, and for someone who typically relies on NPM packages and Gulp, I'm finding it challenging to get up in running with the Hugo SCSS Pipe.

As many have mentioned on the forums, we are stuck once we see "You need dart-sass-embedded in your system $PATH." I've been trying to understand what this means and have made a little progress. It appears $PATH is the location of your /usr/local/bin on a Mac. When viewing this in Finder, I see what looks like other applications that have been installed by Brew.

The problem I'm having is trying to install https://github.com/sass/dart-sass-embedded as mentioned in the error. I don't know how to get Dart Sass Embedded into my system $PATH. Typically, I would rely on Brew or NPM to install dependencies. All the Google searches describe things related to Netlify or some other host, but I’m not using any of those. I’m simply trying to figure out how to get it running locally so I can build and then upload my files to my host via FTP.

Any thoughts on how to install Dart Sass Embedded into my system $PATH so I can run that Hugo Pipe? I appreciate any advice you can give. Thanks

**Reply by brycewray** [reply]
Thanks for the kind words!

Long-time Hugo regular Zach Betz solves the `$PATH` conundrum for you in his article, “[How to Add a Binary (or Executable, or Program) to Your PATH on macOS, Linux, or Windows](https://zwbetz.com/how-to-add-a-binary-to-your-path-on-macos-linux-windows/).” Use his procedure to create a new, easily accessed folder in the `$PATH`; then, move the appropriate `dart-sass-embedded` binary file to that new folder. Give that a try and let me know if you need more help with this.

Since you mentioned Homebrew: you can also use this newly created `$PATH` addition to install each new Hugo release immediately after it’s available on GitHub, rather than waiting for it to become available on Homebrew. **However**, you’ll then have to play footsy with the Mac **Security and Privacy** System Preferences, because the Mac initially regards the `hugo` binary as unsafe when you try using it the first time — so it’s up to you whether you’re willing to go through that every time. It’s worth it to **me**, because I’m impatient. 😄

**Reply by burtcrismore** [reply]
Thank you for the quick reply. I've been looking over Zach's article and will give that a try for folder creation. Would I create a new folder inside here /usr/local/bin or somewhere else? Also, which bin file(s) would I place there from the git repo? Would it be this https://github.com/sass/dart-sass-embedded/tree/main/bin#:~:text=dart_sass_embedded.dart or the whole repo inside this new folder? Thanks

**Reply by brycewray** [reply]
**UPDATE, 2022-05-20, 19:07CDT**: I am editing this later after learning that the procedure I originally outlined here, although it somehow **had** been working for me, isn’t correct after all.

I’d just follow Zach’s procedure exactly (note that it requires identifying your shell config file, either `.zshrc` or `.bashrc`). That way, your usual directory will get the `${HOME}/bin` folder and it’ll always be easily visible in the Mac Finder for simple drag-and-drop. **You’ll also want to add `${HOME}/bin/sass_embedded` to the `$PATH`.** To get Embedded Dart Sass itself, go to the [Embedded Dart Sass releases page](https://github.com/sass/dart-sass-embedded/releases) and: 

1. Download the correct `.tar.gz` archive file (either `arm64` for an Apple Silicon Mac or `x64` for an Intel Mac).
2. Unpack it by double-clicking to use the Mac’s built-in capability. The result will be a `sass_embedded` folder.
3. Move that entire folder into your new `bin` folder. (Replace any existing `sass_embedded` folder you may already have there.)

**By the way**: always check the version to make sure it’s what it should be:

```bash
dart-sass-embedded --version
```

For v.1.52.1, for example, you should get:

```bash
{
  "protocolVersion": "1.0.0",
  "compilerVersion": "1.52.1",
  "implementationVersion": "1.52.1",
  "implementationName": "Dart Sass",
  "id": 0
}
```

**Reply by brycewray** [reply]
@burtcrismore Pinging you to make sure you saw my update, above. Hope this helps.

**Reply by burtcrismore** [reply]
Hey Bryce, thanks so much for the thoughtful reply. This really is helpful. I've learned a lot with your examples. I'm getting closer to a resolution, but I think I missed something. I did follow Zach's instructions to the T, or so I think. 

I created a bin folder at `$HOME` and can see it in Finder. I did drop the binary file `dart-sass-embedded`, however, when I run `--version` it says `bin/src/dart: No such file or directory`. When I run `which dart-sass-embedded` it shows it at `bin/dart-sass-embedded` and not at the `sass_embedded` you described. I think in following Zach's instruction and his use of the generic `binary` I thought that was just the file. It sounds like I should create another folder called `sass_embedded` correct? I'm assuming I'll need to run chmod 755 on that whole directory as well?

One good thing is that when I ran `hugo server -D` it is now displaying a different message. `unexpected EOF when executing "dart-sass-embedded". The user running hugo must have read and execute permissions on this program.`

Thanks again for you help. I really do appreciate you taking the time.

**Reply by burtcrismore** [reply]
Interesting, when I added the `sass_embedded` folder to `${HOME}/bin` it throws that error again that `dart-sass-embedded` needs to be in the system path. I guess I'll try and place the file back in the `${HOME}/bin` directly.

**Reply by brycewray** [reply]
@burtcrismore No, actually, it sounds as if, unfortunately, you may have followed my earlier (apparently but not intentionally wrong) instructions. Turns out that (a.) you have to move the **whole** `sass_embedded` folder (from that `.tar.gz` archive) into the `{$HOME}/bin` folder **and** (b.) you also have to add that folder — thus, `{$HOME}/bin/sass_embedded` — to your `$PATH`. Part (b.) is where your **`${HOME}/.zshrc`** file will come in, as per the [other page to which Zach linked](https://zwbetz.com/shell-config-file-on-mac/). So add the following lines to your `${HOME}/.zshrc` file to append **both** `{$HOME}/bin` and `{$HOME}/bin/sass_embedded` to your `$PATH`:

```bash
export PATH="$HOME/bin:$PATH"
export PATH="$HOME/bin/sass_embedded:$PATH"
```
... and then you should be able to ask about the version and get a properly working reply:

```bash
dart-sass-embedded --version
```

(In all of the above examples, the `${HOME}` variable represents your normal user directory — perhaps `/Users/BurtCrismore` or something like that.)

What I originally told you **had** been working for me, weirdly enough, but — as I found out today when I tried installing the 1.52.1 version of Dart Sass Embedded — I’d apparently only been lucky and it shouldn’t have worked. Again, my apologies for any time you’ve wasted due to what I said earlier before I edited it out. 😞

**Reply by burtcrismore** [reply]
No worries at all. I just appreciate you taking the time help. Your instructions worked perfectly and were easy to understand. I am now able to use Dart Sass and I learned about adding binary files to the path in the process. Thank you so much. Your instructions should be part of the Hugo documentation.

I assume whenever there is a new version I would delete the old copy and place the new one in its place?  Do I need to make the binary file executable again or is that no longer necessary?

**Reply by brycewray** [reply]
> No worries at all. I just appreciate you taking the time help. Your instructions worked perfectly and were easy to understand. I am now able to use Dart Sass and I learned about adding binary files to the path in the process. Thank you so much. Your instructions should be part of the Hugo documentation.

Well, I wouldn't go that far, but I am glad you’re good to go!
 
> I assume whenever there is a new version I would delete the old copy and place the new one in its place? Do I need to make the binary file executable again or is that no longer necessary?

When you drag over a new `sass_embedded` folder, it should prompt you whether you want to replace the old one. If not, however, yes, you **do** want to get rid of the old one. And the `${HOME}/bin/sass_embedded` entry is already in your `$PATH` as a result of today’s activities, so that should be fine from there on. (Only way it **wouldn’t** be is if, for some reason, the Sass folks change the name of the folder — but, if by some bizarre chance that happens, just follow the aforementioned instructions to append the **new** folder name to your `${PATH}`.)

**Reply by brycewray** [reply]
I will further add, for the sake of future visitors seeing this, that where I slipped up originally in the thread above was in thinking that `sass_embedded/dart-sass-embedded` was the binary (because that’s what it appeared to be in the Mac Finder); but, after doing research on Linux and Windows, I found that it was in fact just a **shell script** to run the **real** binary, which is `sass_embedded/src/dart`. Arggh. Sorry! I will make appropriate corrections wherever they may be required.
